### PR TITLE
Feat/eng 485 rebond

### PR DESCRIPTION
**Pull request by gjgd**
https://transmute.atlassian.net/browse/ENG-485

### CODE REVIEW

File: contracts/TransmuteDPOS.sol [hunk 1]
```diff
@@ -76,11 +65,13 @@ contract TransmuteDPOS is TransmuteToken, RoundManager, DelegatorManager {
...
-    withdrawInformations[msg.sender] = WithdrawInformation(withdrawBlock, amount);
+    unbondingInformations[msg.sender] = UnbondingInformation(withdrawBlock, amount);
     if (d.delegateAddress == msg.sender) {
-      // A Provider has to be a Delegator to himself
-      // Therefore if a Provider unbonds he should resign
-      resignAsProvider(msg.sender);
+      if (providerStatus(msg.sender) == ProviderStatus.Registered) {
+        // A Provider has to be a Delegator to himself
+        // Therefore if a Provider unbonds he should resign
+        resignAsProvider(msg.sender);
```
Comment (OR13) [inline contracts/TransmuteDPOS.sol]: I wonder about the logic of this. It could be the case that we throw here, and require the provider to call resign directly. interpreting the user intention seems worse then forcing them to follow a specific order of operations.... is there are counter point?

File: contracts/TransmuteDPOS.sol [hunk 2]
```diff
@@ -93,14 +84,41 @@ contract TransmuteDPOS is TransmuteToken, RoundManager, DelegatorManager {
...
-    this.transfer(msg.sender, withdrawInformation.amount);
-    delete withdrawInformations[msg.sender];
+    UnbondingInformation storage unbondingInformation = unbondingInformations[msg.sender];
+    require(unbondingInformation.withdrawBlock <= block.number);
+    require(delegatorStatus(msg.sender) == DelegatorStatus.Unbonding);
+    this.transfer(msg.sender, unbondingInformation.amount);
+    delete unbondingInformations[msg.sender];
+  }
+
+  function rebond(address _provider) external {
```
Comment (OR13) [inline contracts/TransmuteDPOS.sol]: why is rebond necessary, why not just have `bond` ?

File: test/unit/TransmuteDPOS.spec.js [hunk 3]
```diff
@@ -506,6 +515,135 @@ contract('TransmuteDPOS', (accounts) => {
...
+      assert.equal(DELEGATOR_UNBONDING, await tdpos.delegatorStatus(delegator2));
+      const unbondingInformation = await tdpos.unbondingInformations.call(delegator2);
+      const previousAmountBonded = unbondingInformation[1];
+      await tdpos.rebond(provider2, {from: delegator2});
+      const delegator = await tdpos.delegators.call(delegator2);
+      const amountBonded = delegator[1];
+      assert.deepEqual(previousAmountBonded, amountBonded);
+    });
+
+    it('should fail if target is not a Registered Provider or the Delegator himself', async () => {
```
Comment (OR13) [inline test/unit/TransmuteDPOS.spec.js]: I think this test could be split up into 3 tests.

File: contracts/TransmuteDPOS.sol (hunk 2 above)
Comment (gjgd) [reply contracts/TransmuteDPOS.sol]: Good question, I did a first iteration bond() and rebond were merged into the same function but ran into problems if the amount you choose to rebond() is not the same the amount you bonded initially.

Example:
- D bond 10 TST to P
- D unbonds, so an unbondingInformation object is created containing "withdrawBlock" and "amount" information, that can be used for `withdraw()` or for `rebond()`
- if D rebonds less than 10 TST, let's say 6 there is a problem. It puts D in an in between state where he is Bonded to a new Provider but he still Unbonding because he has unbondingInformation (the 4 TST left). Moreover this new unbondingInformation object can't be treated like the previous one because he can only be `withdrawn()` not `rebonded()` (or else D would bond to two different providers which is against the rules).
- D rebonds exactly 10 TST then D is Bonded to the new Provider and unbondingInformation is deleted.

In short, it adds a lot less complexity if rebond() does not take an `amount` parameter and just automatically rebonds the full amount that has been unbonded.

File: contracts/TransmuteDPOS.sol (hunk 1 above)
Comment (gjgd) [reply contracts/TransmuteDPOS.sol]: If a Provider unbonds, he breaks rule number 1 of being a Provider: "A Provider must bond to himself before registering". Therefore if he unbonds, he is by definition not a Provider anymore (or else all the code that relies on rule number 1 would break).

I agree that my comment is misleading. It's not like he should resign and the contract is making the decision for him. It's that he isn't a Provider anymore.

File: contracts/TransmuteDPOS.sol (hunk 2 above)
Comment (OR13) [reply contracts/TransmuteDPOS.sol]: this makes sense.
